### PR TITLE
[ENHANCEMENT] Initial implementation of css hot realod

### PR DESCRIPTION
**Pull request by gcollazo**
[Link to original PR](https://github.com/ember-cli/ember-cli/pull/5162)

Replaces https://github.com/ember-cli/ember-cli/issues/2371.

### CODE REVIEW

File: lib/tasks/server/livereload-server.js [hunk 1]
```diff
@@ -81,6 +85,15 @@ module.exports = Task.extend({
 
   didChange: function(results) {
     var filePath = path.relative(this.project.root, results.filePath || '');
+    var appStyleResource = this.project.pkg.name + '.css';
+
+    var reloadPaths = function() {
```
Comment (stefanpenner) [inline lib/tasks/server/livereload-server.js]: i believe the function can be removed, and rather reloadPaths calculated here directly.

File: lib/tasks/server/livereload-server.js [hunk 2]
```diff
@@ -81,6 +85,15 @@ module.exports = Task.extend({
...
+      if (reloadCssPattern.test(filePath)) {
+        return [appStyleResource];
```
Comment (stefanpenner) [inline lib/tasks/server/livereload-server.js]: what if there are other css files that changes?

**Comment by stefanpenner** [comment]
This seems to be taking the FS events as a cue to figure out what actually changed. That may work sometimes, but not always.

**Comment by stefanpenner** [comment]
I have submitted a PR to your branch (which is this PR) with my recommendations https://github.com/gcollazo/ember-cli/pull/1

**Comment by stefanpenner** [comment]
It appears that hot-reloading with css that has sourcemaps (like sass, the sourcemaps dont appear to reload correctly).... I tried both embedded and not embedded but without luck. Thats not a blocker for getting this in, but worth some investigation.

**Comment by gcollazo** [comment]
@stefanpenner Just pushed a few changes. Most noticeable things are:

- Filter out *.map files to prevent a full browser reload when a precompiled style (scss, styl, less) file changes
- Filter out directories since live reload can't do anything useful with them. This also helps with tests.
- Removed all references to `Project.liveRealoadFilterPatterns` and it' related tests since I couldn't find any references to that on the project.

Let me know how can I improve it.

**Comment by rwjblue** [comment]
liveRealoadFilterPatterns Is used by addons

**Comment by stefanpenner** [comment]
@rwjblue for what?

**Comment by rwjblue** [comment]
Suppressing live reload for certain files. 

Robert Jackson

-- twitter: rwjblue
-- github: rjackson

**Comment by stefanpenner** [comment]
@rwjblue seems backwards the watchers filters should be used to prevent builds. Or is there a use-case where a built should happen, but a live reload shouldn't. That seems dubious.

I suspect anyone doing this, is likely doing so to prevent spurious live reloads, due to spurious builds, likely due to various temp files from various editors being written. These scenarios should actually be guarded at the watcher.

I am unaware of a reasonable use-case, where a build should be triggered, but a life reload should be prevented. If a use-case exists, please share.

**Comment by rwjblue** [comment]
@stefanpenner - I am not really saying that it should be kept, just responding to the comment that "X isn't used internally so we should remove it".

The original use case for `Project.liveReloadFilterPatterns`'s was for the ember-cli-proxy-fixtures addon to use it as a filter to enable "VCR" like proxy fixture recording during a test suite run. On first run, the addon writes AJAX request fixtures to the `tests/fixtures` directory, which would cause a livereload (and restart the test suite).  The filter pattern was created/used to prevent that build from triggering a livereload.  This seems like a reasonable use case for wanting a build (so that subsequent refreshes include the fixtures properly) but prevent livereload.  However, if we want to remove the functionality that seems fine as long as it is intended for a major version and the breakage in API is documented when we release.

**Comment by stefanpenner** [comment]
That sounds like a legit use-case, @gcollazo mind adding back that support? (I realized I removed it, so if you don't have time I can likely doit later)

**Comment by gcollazo** [comment]
I'll do it.

**Comment by stefanpenner** [comment]
@gcollazo be sure to use the `files` as the source of truth for what changed.

**Comment by gcollazo** [comment]
@stefanpenner didn't read that last message before making these changes. What I did there should be backwards compatible. I'll give it another shot later tonight.

File: lib/tasks/server/livereload-server.js [hunk 3]
```diff
@@ -76,21 +100,39 @@ module.exports = Task.extend({
...
+  getDirectoryEntries: function(directory) {
+    return walkSync.entries(directory);
   },
 
   didChange: function(results) {
-    var filePath = path.relative(this.project.root, results.filePath || '');
+    var canTrigger = true;
+
+    if (this.project.liveReloadFilterPatterns.length > 0) {
+      var filePath = path.relative(this.project.root, results.filePath || '');
```
Comment (stefanpenner) [inline lib/tasks/server/livereload-server.js]: can we extract this block into a separate function, it will make it easer to read/test etc.

Comment (stefanpenner) [inline lib/tasks/server/livereload-server.js]: I believe `canTrigger` may be flawed, as the filePath isn't going to always be right. It is very easy for many files to change, and get batched together...

**Comment by gcollazo** [comment]
@stefanpenner @rwjblue 

If you look at [ember-cli-proxy-fixtures](https://github.com/dockyard/ember-cli-proxy-fixtures/blob/master/index.js#L61) they are filtering based on the path of the input file (tests/fixtures/proxy/file-a.js) which makes sense from the perspective of the user.

Right now `previousTree.calculatePatch(this.tree)` gives us a diff between trees but it seems there's only information about the destination of files (assets/tests.js) and not about their source.

In orden to really maintain backwards compatibility we need to know the original path of the files. The only way that I know of to achieve this is using `filePath` which only reports one file causing us to potentially miss files. Suggestions?

I searched all addons on npm and ember-cli-proxy-fixtures and ember-cli-styles-reloader are the only addons making reference to `liveReloadFilterPatterns`. Maybe we should just ask if the current implementation works for them and keep that. Thoughts?

**Comment by stefanpenner** [comment]
> In orden to really maintain backwards compatibility we need to know the original path of the files. The only way that I know of to achieve this is using filePath which only reports one file causing us to potentially miss files. Suggestions?

seems like we just keep that old behavior for now, rethink again soon.

**Comment by gcollazo** [comment]
@stefanpenner moved the skip reload logic to `shouldTriggerReload` and added a few more tests.

**Comment by gcollazo** [comment]
@stefanpenner earlier today @rwjblue confirmed that the test failures are not related to this PR. I think this is ready for review whenever you have the time.

**Comment by gcollazo** [comment]
*Adding this info here just to keep track of it:*

I did some testing based on [this tweet](https://twitter.com/addyosmani/status/671378820021264384) to try to determine why css source maps where not updating after live reload and it seems like an issue with chrome. I made some videos and demo apps with ember-cli and webpack.

Tested on Chrome:

- 47.0.2526.73 (64-bit): Has the problem
- 47.0.2526.73 beta (64-bit): Has the problem
- 49.0.2579.0 canary (64-bit): Works sometimes

Using ember-cli on Beta:
 - Video [https://youtu.be/P2_Te5LdXYE](https://youtu.be/P2_Te5LdXYE)
 - [Demo app source](https://github.com/gcollazo/ember-cli-livereload-source-maps-demo)

Using webpack on Beta:
- Video [https://youtu.be/X0fLP3tyNNQ](https://youtu.be/X0fLP3tyNNQ)
- [Demo app source](https://github.com/gcollazo/-webpack-livereload-source-maps-demo)

Using webpack on Canary:
- Video [https://youtu.be/8UmcadkpkFA](https://youtu.be/8UmcadkpkFA)

**Comment by stefanpenner** [comment]
@gcollazo awesome, thanks for the investigation. I suspect, this should be reported [here](https://code.google.com/p/chromium/issues/list?can=2&q=Cr%3DPlatform-DevTools)

**Comment by stefanpenner** [comment]
@gcollazo this looks really good, thanks for iterating with me on this one. I'm going to take at the (very likely to be unrelated) test failures shortly and try and get them green so we can merge

**Comment by rwjblue** [comment]
@homu r+

**Comment by rwjblue** [comment]
@gcollazo - Thanks for all your hard work on this!

**Comment by gcollazo** [comment]
Thanks @rwjblue and @stefanpenner for the help.

**Comment by cibernox** [comment]
This. Is. Big. :tada:

**Comment by raido** [comment]
Awesome 🎉

**Comment by matthooks** [comment]
Just wanted to leave props for doing this work! :+1:

**Comment by lukemelia** [comment]
Great job!

**Comment by gfvcastro** [comment]
Awesome :tada: :+1:

**Comment by ynnoj** [comment]
:clap:

**Comment by he9qi** [comment]
:+1: to everyone involved. Great job!!!

**Comment by sandstrom** [comment]
Awesome! Thanks! :balloon:

**Comment by bradcrawford** [comment]
👍

**Pull request merged**